0012/README.md

```diff
@@ -0,0 +1,233 @@
+  For example, `resourceID` will be set to `workflowID` for workflow resource and to `templateID` for the template resource.
+- `resouceType`: type of the resource (template, hardware, workflow) for which the event was generated
+- `eventType`: the event verb; describing the action on resource that generated the event
+- `time`: the event timestamp
+- `data`: the primary event payload, represented as `interface{}`.
+  For a workflow created event, the payload (data) can be the complete workflow structure.
+  For an on-timeout event the data can be an action that needs to be executed next.
+
+### EventClient
+
+A new `EventClient` should be developed with the primitive required for the events at least `Watch`.
+At the beginning, only a `Watch` function is required but we think a natural evolution will be to serve a function that can be used from other components to fire events.
+Or maybe we can do a `Watch` and `Create` straight away.
+All the resource clients will have a `Watch` function that will stream events for that particular resource.
+
+```
+Watch(ctx context.Context, in EventWatchRequest)
+```
+
+Where the `EventWatchRequest` can take the following structure:
+
+```
+type EventWatchRequest {
+    EventName  // the name of the event to filter by
+    ResourceID // the resource ID (ResourceType required when this is set)
+    ResourceType // workflow, template, action, hardware
+}
+```
+
+So when we create a `HardwareClient`, for instance, it would return a watcher for the `Hardware` resource type.
+All the clients like `HardwareClient` will be using `EventClient` under the hood.
+
+There is no logic to keep track of which events are fired or sent to a consumer.
+Every consumer when it connects to a stream of events will specify how old the events returned should be (by default 5m).
+In order to make it possible, all the events will be stored in the Events table.
+The Events table will (roughly) have the following structure:
+
+```
+tinkerbell=# \d events
+                           Table "public.events"
+    Column     |           Type           | Collation | Nullable | Default
+---------------+--------------------------+-----------+----------+---------
+ id            | uuid                     |           | not null |
+ resource_id   | uuid                     |           | not null |
+ resource_type | integer                  |           | not null |
+ event_type    | integer                  |           | not null |
+ created_at    | timestamp with time zone |           |          |
+ data          | jsonb                    |           |          |
+Indexes:
+    "events_pkey" PRIMARY KEY, btree (id)
+```
+
+The tink-server will only be responsible for generating the events.
+It will not contain any business logic that needs to be executed as an event occurs.
+Instead it's the consumer who will have to implement the business logic in an idempotent way.
```

Comment:
how are we handling the scenario where an event was triggered but consumer didn't received it ?
In that case, neither consumer is aware if event was triggered nor tink is aware if event was received ?
Also, a possible of re-delivery.

Reply:
Adding a delivery/retry logic will be an overkill at this point. Every consumer when it connects to a stream of events will specify how old the events returned should be (by default 5m). That way they will receive the events missed during the connect breakdown.

Reply:
I don't agree it would be an overkill. 

Consumer is unaware if an event was triggered. Tink will assume it should have been received by the consumer. What if consumer never asks or when it asks, then some events are left out for not being the part of the `time-period`. Consumer will endup being inconsistent and can't be blamed since it is unaware if an event was triggered.

Reply:
> how are we handling the scenario where an event was triggered but consumer didn't received it ?

We do not have a catalog of consumers. Consumers do not register themself (consumers are normal clients like the tink-cli) if a consumer is not attached to a stream or if it didn't ask for old enough events for the tink-server that consumer does not exist. When a consumer gets implemented it knows that tink-server stores events for some amount of time, it is responsibility of who implements the consumer to figure out how to be reliable. 

The tink-server responsibility is to store and stream events reliably to who ever asks for them.

Reply:
I think my comment was deleted. Let me add it again. 

Event is streamed to consumer when it occurs and also, consumer is listening onto the stream and is unaware about the occurrence of an event until it receives it.

> The tink-server responsibility is to store and stream events reliably and fastly to whoever asks for them

How are we ensuring the streaming events reliably, what happens when the event is lost during the stream, may be due to network issue or service restart? Just like with a REST call we are aware about the status of the execution with status-code and resend it or take corrective actions.

Reply:
I think this is answered in https://github.com/tinkerbell/proposals/pull/17/#discussion_r488916179